File: frontend/src/utils/displayName.test.ts
import { describe, it, expect } from 'vitest';
import { normalizeDisplayName, isProviderNameValid, isChannelNameValid } from './displayName';

// 不可见字符用显式 \u 转义：BOM=\ufeff  NEL=\u0085  ZWSP=\u200b  RLO=\u202e

describe('normalizeDisplayName', () => {
  it('strips edge whitespace', () => expect(normalizeDisplayName('  赛博AI  ')).toBe('赛博AI'));
  it('strips edge BOM', () => expect(normalizeDisplayName('\ufeff赛博AI\ufeff')).toBe('赛博AI'));
  it('strips edge NEL', () => expect(normalizeDisplayName('\u0085Sai\u0085')).toBe('Sai'));
  it('strips edge ZWSP', () => expect(normalizeDisplayName('\u200b赛博AI\u200b')).toBe('赛博AI'));
  it('keeps interior invisible (rejected later, not stripped)', () =>
    expect(normalizeDisplayName('Sai\u200bAI')).toBe('Sai\u200bAI'));
  it('pure invisible -> empty', () => expect(normalizeDisplayName('\u200b\ufeff')).toBe(''));
});

describe('isProviderNameValid', () => {
  it('accepts chinese', () => expect(isProviderNameValid('赛博AI')).toBe(true));
  it('accepts edge BOM (stripped)', () => expect(isProviderNameValid('\ufeff赛博AI')).toBe(true));
  it('accepts edge NEL (stripped)', () => expect(isProviderNameValid('\u0085Sai AI')).toBe(true));
  it('rejects interior zero width', () => expect(isProviderNameValid('Sai\u200bAI')).toBe(false));
  it('rejects interior bidi', () => expect(isProviderNameValid('Sai\u202eAI')).toBe(false));
  it('rejects empty', () => expect(isProviderNameValid('')).toBe(false));
  it('rejects pure invisible', () => expect(isProviderNameValid('\u200b\ufeff')).toBe(false));
  it('rejects over 100 runes', () => expect(isProviderNameValid('赛'.repeat(101))).toBe(false));
  it('accepts exactly 100 runes', () => expect(isProviderNameValid('赛'.repeat(100))).toBe(true));
});

describe('isChannelNameValid', () => {
  it('accepts empty (optional)', () => expect(isChannelNameValid('')).toBe(true));
  it('accepts pure invisible -> empty', () => expect(isChannelNameValid('\u200b')).toBe(true));
  it('accepts chinese', () => expect(isChannelNameValid('华东线路')).toBe(true));
  it('rejects interior bidi', () => expect(isChannelNameValid('线\u202e路')).toBe(false));
  it('rejects interior zero width', () => expect(isChannelNameValid('线\u200b路')).toBe(false));
  it('rejects over 40 runes', () => expect(isChannelNameValid('赛'.repeat(41))).toBe(false));
  it('accepts exactly 40 runes', () => expect(isChannelNameValid('赛'.repeat(40))).toBe(true));
});

File: internal/displayname/displayname_test.go
package displayname

import (
	"strings"
	"testing"
)

// 不可见字符全部用显式 \u 转义，避免源码里字面隐形字符的维护风险与歧义。
//   BOM  = \ufeff (Cf)   NEL = \u0085 (Cc)   ZWSP = \u200b (Cf)
//   RLO  = \u202e (Cf)   Zl  = \u2028        Zp   =

func TestValidateProviderName(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		want    string
		wantErr bool
	}{
		{"ascii unchanged", "SaiAI", "SaiAI", false},
		{"ascii with spaces", "Sai AI Cloud", "Sai AI Cloud", false},
		{"ascii punctuation", "WorldBase.ai", "WorldBase.ai", false},
		{"chinese", "赛博AI", "赛博AI", false},
		{"japanese", "さくらクラウド", "さくらクラウド", false},
		{"russian", "Яндекс", "Яндекс", false},
		{"arabic rtl ok", "سحابة", "سحابة", false},
		{"mixed cjk+ascii", "赛博AI Cloud", "赛博AI Cloud", false},
		{"trims surrounding space", "  赛博AI  ", "赛博AI", false},
		{"strips edge BOM", "\ufeff赛博AI\ufeff", "赛博AI", false},
		{"strips edge NEL", "\u0085Sai AI\u0085", "Sai AI", false},
		{"strips edge ZWSP", "\u200bSai\u200b", "Sai", false},
		{"strips edge RLO then valid", "\u202eSaiAI", "SaiAI", false},
		{"empty", "", "", true},
		{"all whitespace", "   ", "", true},
		{"pure BOM -> empty -> required err", "\ufeff\ufeff", "", true},
		{"interior tab (Cc)", "Sai\tAI", "", true},
		{"interior newline (Cc)", "Sai\nAI", "", true},
		{"interior zero width (Cf)", "Sai\u200bAI", "", true},
		{"interior bidi override (Cf)", "Sai\u202eAI", "", true},
		{"interior line separator (Zl)", "Sai\u2028AI", "", true},
		{"interior paragraph separator (Zp)", "Sai\u2029AI", "", true},
		{"over 100 runes", strings.Repeat("赛", 101), "", true},
		{"exactly 100 runes ok", strings.Repeat("赛", 100), strings.Repeat("赛", 100), false},
		{"invalid utf-8", "Sai\xffAI", "", true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateProviderName(tt.input)
			if (err != nil) != tt.wantErr {
				t.Fatalf("ValidateProviderName(%q) err=%v, wantErr=%v", tt.input, err, tt.wantErr)
			}
			if !tt.wantErr && got != tt.want {
				t.Fatalf("ValidateProviderName(%q)=%q, want %q", tt.input, got, tt.want)
			}
		})
	}
}

func TestValidateChannelName(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		want    string
		wantErr bool
	}{
		{"empty is valid (optional)", "", "", false},
		{"all whitespace -> empty", "   ", "", false},
		{"pure edge Cf -> empty valid", "\u200b\ufeff", "", false},
		{"chinese ok", "线路一", "线路一", false},
		{"trims + strips edge BOM", "\ufeff线路一\ufeff", "线路一", false},
		{"interior zero width rejected", "线\u200b路", "", true},
		{"interior bidi rejected", "线\u202e路", "", true},
		{"over 40 runes", strings.Repeat("赛", 41), "", true},
		{"exactly 40 runes ok", strings.Repeat("赛", 40), strings.Repeat("赛", 40), false},
		{"invalid utf-8", "a\xffb", "", true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateChannelName(tt.input)
			if (err != nil) != tt.wantErr {
				t.Fatalf("ValidateChannelName(%q) err=%v, wantErr=%v", tt.input, err, tt.wantErr)
			}
			if !tt.wantErr && got != tt.want {
				t.Fatalf("ValidateChannelName(%q)=%q, want %q", tt.input, got, tt.want)
			}
		})
	}
}
